File: android/app/src/main/java/xyz/sonet/app/viewmodels/ContentCreationViewModel.kt
package xyz.sonet.app.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import xyz.sonet.app.grpc.SonetGRPCClient
import xyz.sonet.app.grpc.SonetConfiguration
import xyz.sonet.app.grpc.proto.*
import java.util.*

class ContentCreationViewModel(application: Application) : AndroidViewModel(application) {
    
    // MARK: - State Flows
    private val _noteContent = MutableStateFlow("")
    val noteContent: StateFlow<String> = _noteContent.asStateFlow()
    
    private val _selectedMedia = MutableStateFlow<List<MediaItem>>(emptyList())
    val selectedMedia: StateFlow<List<MediaItem>> = _selectedMedia.asStateFlow()
    
    private val _replyToNote = MutableStateFlow<Note?>(null)
    val replyToNote: StateFlow<Note?> = _replyToNote.asStateFlow()
    
    private val _quoteNote = MutableStateFlow<Note?>(null)
    val quoteNote: StateFlow<Note?> = _quoteNote.asStateFlow()
    
    private val _selectedHashtags = MutableStateFlow<List<String>>(emptyList())
    val selectedHashtags: StateFlow<List<String>> = _selectedHashtags.asStateFlow()
    
    private val _selectedMentions = MutableStateFlow<List<String>>(emptyList())
    val selectedMentions: StateFlow<List<String>> = _selectedMentions.asStateFlow()
    
    private val _isPosting = MutableStateFlow(false)
    val isPosting: StateFlow<Boolean> = _isPosting.asStateFlow()
    
    private val _postingError = MutableStateFlow<String?>(null)
    val postingError: StateFlow<String?> = _postingError.asStateFlow()
    
    private val _showHashtagSuggestions = MutableStateFlow(false)
    val showHashtagSuggestions: StateFlow<Boolean> = _showHashtagSuggestions.asStateFlow()
    
    private val _showMentionSuggestions = MutableStateFlow(false)
    val showMentionSuggestions: StateFlow<Boolean> = _showMentionSuggestions.asStateFlow()
    
    private val _hashtagSuggestions = MutableStateFlow<List<String>>(emptyList())
    val hashtagSuggestions: StateFlow<List<String>> = _hashtagSuggestions.asStateFlow()
    
    private val _mentionSuggestions = MutableStateFlow<List<UserProfile>>(emptyList())
    val mentionSuggestions: StateFlow<List<UserProfile>> = _mentionSuggestions.asStateFlow()
    
    private val _currentHashtagQuery = MutableStateFlow("")
    val currentHashtagQuery: StateFlow<String> = _currentHashtagQuery.asStateFlow()
    
    private val _currentMentionQuery = MutableStateFlow("")
    val currentMentionQuery: StateFlow<String> = _currentMentionQuery.asStateFlow()
    
    private val _scheduledDate = MutableStateFlow<Long?>(null)
    val scheduledDate: StateFlow<Long?> = _scheduledDate.asStateFlow()
    
    private val _showScheduling = MutableStateFlow(false)
    val showScheduling: StateFlow<Boolean> = _showScheduling.asStateFlow()
    
    private val _isDraft = MutableStateFlow(false)
    val isDraft: StateFlow<Boolean> = _isDraft.asStateFlow()
    
    private val _draftId = MutableStateFlow<String?>(null)
    val draftId: StateFlow<String?> = _draftId.asStateFlow()
    
    // MARK: - Private Properties
    private val grpcClient = SonetGRPCClient(application, SonetConfiguration.development)
    private var hashtagSearchJob: kotlinx.coroutines.Job? = null
    private var mentionSearchJob: kotlinx.coroutines.Job? = null
    
    // MARK: - Constants
    private val maxCharacterCount = 280
    private val maxMediaCount = 4
    
    // MARK: - Computed Properties
    val characterCount: Int
        get() = _noteContent.value.length
    
    val remainingCharacters: Int
        get() = maxCharacterCount - characterCount
    
    val canPost: Boolean
        get() = _noteContent.value.trim().isNotEmpty() && 
                !_isPosting.value && 
                remainingCharacters >= 0
    
    val characterCountColor: androidx.compose.ui.graphics.Color
        get() = when {
            remainingCharacters < 0 -> androidx.compose.ui.graphics.Color(0xFFDD3333)
            remainingCharacters <= 20 -> androidx.compose.ui.graphics.Color(0xFF888888)
            else -> androidx.compose.ui.graphics.Color(0xFF6B6B6B)
        }
    
    // MARK: - Initialization
    init {
        setupContentMonitoring()
        loadDraftIfNeeded()
    }
    
    // MARK: - Public Methods
    fun setReplyToNote(note: Note?) {
        _replyToNote.value = note
    }
    
    fun setQuoteNote(note: Note?) {
        _quoteNote.value = note
    }
    
    fun updateNoteContent(content: String) {
        _noteContent.value = content
        detectHashtagsAndMentions(content)
    }
    
    fun addMedia(media: MediaItem) {
        if (_selectedMedia.value.size < maxMediaCount) {
            val currentMedia = _selectedMedia.value.toMutableList()
            currentMedia.add(media)
            _selectedMedia.value = currentMedia
        }
    }
    
    fun removeMedia(index: Int) {
        if (index < _selectedMedia.value.size) {
            val currentMedia = _selectedMedia.value.toMutableList()
            currentMedia.removeAt(index)
            _selectedMedia.value = currentMedia
        }
    }
    
    fun addHashtag(hashtag: String) {
        val cleanHashtag = if (hashtag.startsWith("#")) hashtag else "#$hashtag"
        if (!_selectedHashtags.value.contains(cleanHashtag)) {
            val currentHashtags = _selectedHashtags.value.toMutableList()
            currentHashtags.add(cleanHashtag)
            _selectedHashtags.value = currentHashtags
        }
        _showHashtagSuggestions.value = false
        _currentHashtagQuery.value = ""
    }
    
    fun removeHashtag(hashtag: String) {
        val currentHashtags = _selectedHashtags.value.toMutableList()
        currentHashtags.remove(hashtag)
        _selectedHashtags.value = currentHashtags
    }
    
    fun addMention(user: UserProfile) {
        val mention = "@${user.username}"
        if (!_selectedMentions.value.contains(mention)) {
            val currentMentions = _selectedMentions.value.toMutableList()
            currentMentions.add(mention)
            _selectedMentions.value = currentMentions
        }
        _showMentionSuggestions.value = false
        _currentMentionQuery.value = ""
    }
    
    fun removeMention(mention: String) {
        val currentMentions = _selectedMentions.value.toMutableList()
        currentMentions.remove(mention)
        _selectedMentions.value = currentMentions
    }
    
    fun postNote() {
        if (!canPost) return
        
        viewModelScope.launch {
            _isPosting.value = true
            _postingError.value = null
            
            try {
                val content = _noteContent.value.trim()
                
                // Start foreground service and upload media, collect URLs
                try {
                    val ctx = getApplication<Application>()
                    val intent = android.content.Intent(ctx, xyz.sonet.app.notifications.UploadForegroundService::class.java)
                    ctx.startForegroundService(intent)
                } catch (_: Exception) {}
                
                // Upload media and collect URLs
                val uploadedUrls = mutableListOf<String>()
                _selectedMedia.value.forEachIndexed { index, media ->
                    if (media.localUri != null) {
                        try {
                            val uri = android.net.Uri.parse(media.localUri)
                            val bytes = getApplication<Application>().contentResolver.openInputStream(uri)?.use { it.readBytes() } ?: byteArrayOf()
                            val resp = grpcClient.uploadMedia(ownerUserId = SessionViewModel(getApplication()).currentUser.value?.id ?: "current_user", fileName = media.altText ?: "media", mimeType = when (media.type) {
                                MediaType.MEDIA_TYPE_IMAGE -> "image/jpeg"
                                MediaType.MEDIA_TYPE_VIDEO -> "video/mp4"
                                MediaType.MEDIA_TYPE_GIF -> "image/gif"
                                else -> "application/octet-stream"
                            }, bytes = bytes) { progress ->
                                try {
                                    val nm = getApplication<Application>().getSystemService(android.content.Context.NOTIFICATION_SERVICE) as android.app.NotificationManager
                                    val notif = xyz.sonet.app.notifications.MessagingNotificationHelper.buildUploadProgressNotification(
                                        getApplication(),
                                        "Uploading…",
                                        (progress * 100).toInt(),
                                        100,
                                        content = media.altText ?: "Media ${index + 1} of ${_selectedMedia.value.size}"
                                    )
                                    nm.notify(1001, notif)
                                } catch (_: Exception) {}
                            }
                            uploadedUrls.add(resp.url)
                        } catch (e: Exception) {
                            // Fallback to existing url
                            if (media.url.isNotEmpty()) uploadedUrls.add(media.url)
                        }
                    } else if (media.url.isNotEmpty()) {
                        uploadedUrls.add(media.url)
                    }
                }
                
                // Create note request with uploaded URLs
                val noteRequest = CreateNoteRequest.newBuilder()
                    .setContent(content)
                    .addAllMediaUrls(uploadedUrls)
                    .addAllHashtags(_selectedHashtags.value.map { it.removePrefix("#") })
                    .addAllMentions(_selectedMentions.value.map { it.removePrefix("@") })
                    .setVisibility(NoteVisibility.NOTE_VISIBILITY_PUBLIC)
                    .build()
                
                // Add reply info
                _replyToNote.value?.let { note ->
                    noteRequest.replyToNoteId = note.noteId
                }
                
                // Add quote info
                _quoteNote.value?.let { note ->
                    noteRequest.quoteNoteId = note.noteId
                }
                
                // Add scheduling
                _scheduledDate.value?.let { timestamp ->
                    val grpcTimestamp = Timestamp.newBuilder()
                        .setSeconds(timestamp / 1000)
                        .build()
                    noteRequest.scheduledAt = grpcTimestamp
                }
                
                // Post the note
                val response = grpcClient.createNote(noteRequest.build())
                
                if (response.success) {
                    try {
                        val nm = getApplication<Application>().getSystemService(android.content.Context.NOTIFICATION_SERVICE) as android.app.NotificationManager
                        nm.cancel(1001)
                    } catch (_: Exception) {}
                    // Clear form
                    clearForm()
                    
                    // Navigate back or show success
                    // This would be handled by the view
                } else {
                    _postingError.value = response.errorMessage
                }
                
            } catch (error: Exception) {
                _postingError.value = "Failed to post note: ${error.message}"
            } finally {
                _isPosting.value = false
            }
        }
    }
    
    fun saveDraft() {
        val draft = NoteDraft(
            id = _draftId.value ?: UUID.randomUUID().toString(),
            content = _noteContent.value,
            media = _selectedMedia.value,
            replyToNote = _replyToNote.value,
            quoteNote = _quoteNote.value,
            hashtags = _selectedHashtags.value,
            mentions = _selectedMentions.value,
            scheduledDate = _scheduledDate.value,
            createdAt = System.currentTimeMillis()
        )
        
        saveDraftToStorage(draft)
        _draftId.value = draft.id
        _isDraft.value = true
    }
    
    fun loadDraft(draft: NoteDraft) {
        _noteContent.value = draft.content
        _selectedMedia.value = draft.media
        _replyToNote.value = draft.replyToNote
        _quoteNote.value = draft.quoteNote
        _selectedHashtags.value = draft.hashtags
        _selectedMentions.value = draft.mentions
        _scheduledDate.value = draft.scheduledDate
        _draftId.value = draft.id
        _isDraft.value = true
    }
    
    fun clearForm() {
        _noteContent.value = ""
        _selectedMedia.value = emptyList()
        _selectedHashtags.value = emptyList()
        _selectedMentions.value = emptyList()
        _scheduledDate.value = null
        _draftId.value = null
        _isDraft.value = false
        _postingError.value = null
    }
    
    fun schedulePost(date: Long) {
        _scheduledDate.value = date
        _showScheduling.value = false
    }
    
    fun cancelScheduling() {
        _scheduledDate.value = null
        _showScheduling.value = false
    }
    
    fun showScheduling() {
        _showScheduling.value = true
    }
    
    // MARK: - Private Methods
    private fun setupContentMonitoring() {
        // Content monitoring is handled by the detectHashtagsAndMentions function
        // which is called whenever noteContent is updated
    }
    
    private fun detectHashtagsAndMentions(content: String) {
        // Detect hashtags
        val hashtagPattern = Regex("#(\\w+)")
        val hashtagMatches = hashtagPattern.findAll(content)
        
        hashtagMatches.lastOrNull()?.let { match ->
            val query = match.groupValues[1]
            if (query.length >= 2) {
                _currentHashtagQuery.value = query
                searchHashtags(query)
            } else {
                _showHashtagSuggestions.value = false
            }
        } ?: run {
            _showHashtagSuggestions.value = false
        }
        
        // Detect mentions
        val mentionPattern = Regex("@(\\w+)")
        val mentionMatches = mentionPattern.findAll(content)
        
        mentionMatches.lastOrNull()?.let { match ->
            val query = match.groupValues[1]
            if (query.length >= 2) {
                _currentMentionQuery.value = query
                searchUsers(query)
            } else {
                _showMentionSuggestions.value = false
            }
        } ?: run {
            _showMentionSuggestions.value = false
        }
    }
    
    private fun searchHashtags(query: String) {
        hashtagSearchJob?.cancel()
        
        hashtagSearchJob = viewModelScope.launch {
            try {
                val hashtags = grpcClient.searchHashtags(query, 0, 10)
                
                if (!isCancelled) {
                    _hashtagSuggestions.value = hashtags
                    _showHashtagSuggestions.value = true
                }
            } catch (error: Exception) {
                // Handle error silently for hashtag search
            }
        }
    }
    
    private fun searchUsers(query: String) {
        mentionSearchJob?.cancel()
        
        mentionSearchJob = viewModelScope.launch {
            try {
                val users = grpcClient.searchUsers(query, 0, 10)
                
                if (!isCancelled) {
                    _mentionSuggestions.value = users
                    _showMentionSuggestions.value = true
                }
            } catch (error: Exception) {
                // Handle error silently for user search
            }
        }
    }
    
    private fun loadDraftIfNeeded() {
        // Load draft if editing an existing draft
        // This would be implemented based on navigation context
    }
    
    private fun saveDraftToStorage(draft: NoteDraft) {
        // Save draft to SharedPreferences or Room database
        // This would be implemented for draft persistence
    }
}

// MARK: - Data Models
data class MediaItem(
    val id: String = UUID.randomUUID().toString(),
    val mediaId: String = "",
    val url: String = "",
    val type: xyz.sonet.app.grpc.proto.MediaType = xyz.sonet.app.grpc.proto.MediaType.MEDIA_TYPE_IMAGE,
    val altText: String? = null,
    val localUri: String? = null
)

data class NoteDraft(
    val id: String,
    val content: String,
    val media: List<MediaItem>,
    val replyToNote: Note?,
    val quoteNote: Note?,
    val hashtags: List<String>,
    val mentions: List<String>,
    val scheduledDate: Long?,
    val createdAt: Long
)

File: ios/Sonet/Views/ContentCreation/ContentCreationViewModel.swift
import Foundation
import Combine
import SwiftUI
import PhotosUI

@MainActor
class ContentCreationViewModel: ObservableObject {
    
    // MARK: - Published Properties
    @Published var noteContent = ""
    @Published var selectedMedia: [MediaItem] = []
    @Published var replyToNote: Note?
    @Published var quoteNote: Note?
    @Published var selectedHashtags: [String] = []
    @Published var selectedMentions: [String] = []
    @Published var isPosting = false
    @Published var postingError: String?
    @Published var showHashtagSuggestions = false
    @Published var showMentionSuggestions = false
    @Published var hashtagSuggestions: [String] = []
    @Published var mentionSuggestions: [UserProfile] = []
    @Published var currentHashtagQuery = ""
    @Published var currentMentionQuery = ""
    @Published var scheduledDate: Date?
    @Published var showScheduling = false
    @Published var isDraft = false
    @Published var draftId: String?
    
    // MARK: - Private Properties
    private let grpcClient: SonetGRPCClient
    private var cancellables = Set<AnyCancellable>()
    private var hashtagSearchTask: Task<Void, Never>?
    private var mentionSearchTask: Task<Void, Never>?
    
    // MARK: - Constants
    private let maxCharacterCount = 280
    private let maxMediaCount = 4
    
    // MARK: - Computed Properties
    var characterCount: Int {
        noteContent.count
    }
    
    var remainingCharacters: Int {
        maxCharacterCount - characterCount
    }
    
    var canPost: Bool {
        !noteContent.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        !isPosting &&
        remainingCharacters >= 0
    }
    
    var characterCountColor: Color {
        if remainingCharacters < 0 {
            return .red
        } else if remainingCharacters <= 20 {
            return .orange
        } else {
            return .secondary
        }
    }
    
    // MARK: - Initialization
    init(grpcClient: SonetGRPCClient, replyToNote: Note? = nil, quoteNote: Note? = nil) {
        self.grpcClient = grpcClient
        self.replyToNote = replyToNote
        self.quoteNote = quoteNote
        
        setupContentMonitoring()
        loadDraftIfNeeded()
    }
    
    // MARK: - Public Methods
    func addMedia(_ media: MediaItem) {
        guard selectedMedia.count < maxMediaCount else { return }
        selectedMedia.append(media)
    }
    
    func removeMedia(at index: Int) {
        guard index < selectedMedia.count else { return }
        selectedMedia.remove(at: index)
    }
    
    func addHashtag(_ hashtag: String) {
        let cleanHashtag = hashtag.hasPrefix("#") ? hashtag : "#\(hashtag)"
        if !selectedHashtags.contains(cleanHashtag) {
            selectedHashtags.append(cleanHashtag)
        }
        showHashtagSuggestions = false
        currentHashtagQuery = ""
    }
    
    func removeHashtag(_ hashtag: String) {
        selectedHashtags.removeAll { $0 == hashtag }
    }
    
    func addMention(_ user: UserProfile) {
        let mention = "@\(user.username)"
        if !selectedMentions.contains(mention) {
            selectedMentions.append(mention)
        }
        showMentionSuggestions = false
        currentMentionQuery = ""
    }
    
    func removeMention(_ mention: String) {
        selectedMentions.removeAll { $0 == mention }
    }
    
    func postNote() async {
        guard canPost else { return }
        
        isPosting = true
        postingError = nil
        var activity: Any? = nil
        if #available(iOS 16.1, *) {
            let firstFile = selectedMedia.first?.fileName ?? "Uploading"
            activity = UploadActivityManager.start(title: "Posting", filename: firstFile)
        }
        
        do {
            let content = noteContent.trimmingCharacters(in: .whitespacesAndNewlines)
            
            // Upload media first via MediaService.Upload and collect URLs
            var uploadedUrls: [String] = []
            for (index, media) in selectedMedia.enumerated() {
                if let local = media.localImage, let jpeg = local.jpegData(compressionQuality: 0.9) {
                    if #available(iOS 16.1, *), let act = activity as? Activity<UploadActivityAttributes> {
                        UploadActivityManager.update(act, progress: Double(index)/Double(max(selectedMedia.count,1)), filename: media.altText ?? "Image", isCompleted: false)
                    }
                    let upload = try await grpcClient.uploadMedia(ownerId: "current_user", filename: media.altText ?? "image.jpg", mimeType: "image/jpeg", data: jpeg) { p in
                        if #available(iOS 16.1, *), let act = activity as? Activity<UploadActivityAttributes> {
                            UploadActivityManager.update(act, progress: (Double(index) + p)/Double(max(selectedMedia.count,1)), filename: media.altText ?? "Image", isCompleted: false)
                        }
                    }
                    uploadedUrls.append(upload.url)
                } else if !media.url.isEmpty {
                    uploadedUrls.append(media.url)
                }
            }
            // Create note request
            var noteRequest = CreateNoteRequest()
            noteRequest.content = content
            noteRequest.mediaUrls = uploadedUrls
            
            // Add hashtags
            noteRequest.hashtags = selectedHashtags.map { $0.replacingOccurrences(of: "#", with: "") }
            
            // Add mentions
            noteRequest.mentions = selectedMentions.map { $0.replacingOccurrences(of: "@", with: "") }
            
            // Add reply info
            if let replyToNote = replyToNote {
                noteRequest.replyToNoteId = replyToNote.noteId
            }
            
            // Add quote info
            if let quoteNote = quoteNote {
                noteRequest.quoteNoteId = quoteNote.noteId
            }
            
            // Add scheduling
            if let scheduledDate = scheduledDate {
                var timestamp = Timestamp()
                timestamp.seconds = Int64(scheduledDate.timeIntervalSince1970)
                noteRequest.scheduledAt = timestamp
            }
            
            // Set visibility
            noteRequest.visibility = .NOTE_VISIBILITY_PUBLIC
            
            // Post the note
            // Simulate progressive updates if there are media items
            if #available(iOS 16.1, *), let act = activity as? Activity<UploadActivityAttributes> {
                for (index, media) in selectedMedia.enumerated() {
                    let progress = Double(index + 1) / Double(max(selectedMedia.count, 1))
                    UploadActivityManager.update(act, progress: progress, filename: media.fileName ?? "Media", isCompleted: false)
                    try await Task.sleep(nanoseconds: 200_000_000)
                }
            }
            let response = try await grpcClient.createNote(request: noteRequest)
            
            if response.success {
                // Clear form
                clearForm()
                
                // Navigate back or show success
                // This would be handled by the view
                if #available(iOS 16.1, *), let act = activity as? Activity<UploadActivityAttributes> {
                    UploadActivityManager.update(act, progress: 1.0, filename: "Done", isCompleted: true)
                }
            } else {
                postingError = response.errorMessage
            }
            
        } catch {
            postingError = "Failed to post note: \(error.localizedDescription)"
        }
        
        isPosting = false
    }
    
    func saveDraft() {
        let draft = NoteDraft(
            id: draftId ?? UUID().uuidString,
            content: noteContent,
            media: selectedMedia,
            replyToNote: replyToNote,
            quoteNote: quoteNote,
            hashtags: selectedHashtags,
            mentions: selectedMentions,
            scheduledDate: scheduledDate,
            createdAt: Date()
        )
        
        saveDraftToStorage(draft)
        draftId = draft.id
        isDraft = true
    }
    
    func loadDraft(_ draft: NoteDraft) {
        noteContent = draft.content
        selectedMedia = draft.media
        replyToNote = draft.replyToNote
        quoteNote = draft.quoteNote
        selectedHashtags = draft.hashtags
        selectedMentions = draft.mentions
        scheduledDate = draft.scheduledDate
        draftId = draft.id
        isDraft = true
    }
    
    func clearForm() {
        noteContent = ""
        selectedMedia = []
        selectedHashtags = []
        selectedMentions = []
        scheduledDate = nil
        draftId = nil
        isDraft = false
        postingError = nil
    }
    
    func schedulePost(for date: Date) {
        scheduledDate = date
        showScheduling = false
    }
    
    func cancelScheduling() {
        scheduledDate = nil
        showScheduling = false
    }
    
    // MARK: - Private Methods
    private func setupContentMonitoring() {
        // Monitor content changes for hashtag and mention detection
        $noteContent
            .debounce(for: .milliseconds(300), scheduler: DispatchQueue.main)
            .sink { [weak self] content in
                self?.detectHashtagsAndMentions(in: content)
            }
            .store(in: &cancellables)
    }
    
    private func detectHashtagsAndMentions(in content: String) {
        // Detect hashtags
        let hashtagPattern = #"#(\w+)"#
        let hashtagMatches = content.matches(of: hashtagPattern)
        
        if let lastMatch = hashtagMatches.last {
            let query = String(lastMatch.1)
            if query.count >= 2 {
                currentHashtagQuery = query
                searchHashtags(query: query)
            } else {
                showHashtagSuggestions = false
            }
        } else {
            showHashtagSuggestions = false
        }
        
        // Detect mentions
        let mentionPattern = #"@(\w+)"#
        let mentionMatches = content.matches(of: mentionPattern)
        
        if let lastMatch = mentionMatches.last {
            let query = String(lastMatch.1)
            if query.count >= 2 {
                currentMentionQuery = query
                searchUsers(query: query)
            } else {
                showMentionSuggestions = false
            }
        } else {
            showMentionSuggestions = false
        }
    }
    
    private func searchHashtags(query: String) {
        hashtagSearchTask?.cancel()
        
        hashtagSearchTask = Task {
            do {
                let hashtags = try await grpcClient.searchHashtags(query: query, page: 0, pageSize: 10)
                
                if !Task.isCancelled {
                    await MainActor.run {
                        hashtagSuggestions = hashtags
                        showHashtagSuggestions = true
                    }
                }
            } catch {
                // Handle error silently for hashtag search
            }
        }
    }
    
    private func searchUsers(query: String) {
        mentionSearchTask?.cancel()
        
        mentionSearchTask = Task {
            do {
                let users = try await grpcClient.searchUsers(query: query, page: 0, pageSize: 10)
                
                if !Task.isCancelled {
                    await MainActor.run {
                        mentionSuggestions = users
                        showMentionSuggestions = true
                    }
                }
            } catch {
                // Handle error silently for user search
            }
        }
    }
    
    private func loadDraftIfNeeded() {
        // Load draft if editing an existing draft
        // This would be implemented based on navigation context
    }
    
    private func saveDraftToStorage(_ draft: NoteDraft) {
        // Save draft to UserDefaults or Core Data
        // This would be implemented for draft persistence
    }
}

// MARK: - Data Models
struct MediaItem: Identifiable {
    let id = UUID()
    let mediaId: String
    let url: String
    let type: MediaType
    let altText: String?
    let localImage: UIImage?
    
    init(mediaId: String = "", url: String = "", type: MediaType = .MEDIA_TYPE_IMAGE, altText: String? = nil, localImage: UIImage? = nil) {
        self.mediaId = mediaId
        self.url = url
        self.type = type
        self.altText = altText
        self.localImage = localImage
    }
}

struct NoteDraft: Identifiable, Codable {
    let id: String
    let content: String
    let media: [MediaItem]
    let replyToNote: Note?
    let quoteNote: Note?
    let hashtags: [String]
    let mentions: [String]
    let scheduledDate: Date?
    let createdAt: Date
}

// MARK: - Media Types
enum MediaType: String, CaseIterable {
    case image = "image"
    case video = "video"
    case gif = "gif"
    case audio = "audio"
    
    var icon: String {
        switch self {
        case .image: return "photo"
        case .video: return "video"
        case .gif: return "play.rectangle"
        case .audio: return "waveform"
        }
    }
    
    var grpcType: xyz.sonet.app.grpc.proto.MediaType {
        switch self {
        case .image: return .MEDIA_TYPE_IMAGE
        case .video: return .MEDIA_TYPE_VIDEO
        case .gif: return .MEDIA_TYPE_GIF
        case .audio: return .MEDIA_TYPE_AUDIO
        }
    }
}